### ISSUE TITLE
Kodi fails to start due to false-positive detecting if a window manager is running

### ISSUE THREAD

**Issue by bacon-cheeseburger** [xbmc/xbmc#18595]
## Bug report
CWinSystemGbm::InitWindowSystem() checks to see if the DISPLAY or WAYLAND_DISPLAY environment variables exist. If it does, it assumes and windows manager is running and exits. The problem is the existence of either of those does not automatically mean x11 or wayland is running so making that assumption can easily create a false-positive.

## Expected Behavior
If a window manager isn't _actually_ running, I expect Kodi to start.

## Actual Behavior
Kodi exits instead.

## Possible Fix
Check the process list instead of using environment variables to see if a window manager is running? Or some other certain method?

### To Reproduce
1. export DISPLAY=coke or WAYLAND_DISPLAY=pepsi
2. Try to start kodi.
3. Be sad because it exits. :(

### Debuglog
https://paste.kodi.tv/umeboqegir.kodi

### Your Environment
Linux Debian Sid, stable kernel 5.8.12, Kodi compiled from git 19.0-ALPHA2-git+654c1471a6

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by fritsch** [comment]
Come on. 

An X11 application can only start if DISPLAY is properly exported.
A wayland application can only start properly if WAYLAND_DISPLAY is there.

So, what do you expect? Wild guessing? That's the reason why exporting those is a standard requirement of every X11 / Wayland App.

**Comment by bacon-cheeseburger** [comment]
Kodi doesn't know and shouldn't care if a window manager can start, only if one if running because that's the only condition where there is a problem. Maybe the window manager is in a broken state and that's why they're temporarily using gbm. Or maybe they test different setups like many of us do. Environment variables don't tell you what's running so why would you make assumptions when you can check to see what's _actually_ running?

**Comment by fritsch** [comment]
How do you intend to check if an XServer is running?

**Comment by basilgello** [comment]
> How do you intend to check if an XServer is running?

Um! How about getting a window size or screen geometry as a smoke test?

**Comment by fritsch** [comment]
So, you want to write an implementation that does a "test connect" for X11 and Wayland (and others), because one user wrongly exports a DISPLAY environment variable in a non-graphical terminal?

According to the standards: if DISPLAY is exported (X11 Display might be available, which is why all X11 applications will directly connect), if WAYLAND_DISPLAY is available, wayland server might be accessible.

And now it comes: If a windowing system needs a tty text mode terminal only, one pre condition is that neither DISPLAY nor WAYLAND_SYSTEM is exposed. That's how it works on Linux and is standard. You can easily verify that: open a terminal in your graphical desktop session: export |grep -i display -> it's set. If you login after ctl alt F1 and issue export |grep -i display again, it returns empty. If it's exposed - something is wrong as DISPLAY is set by X11 environment and WAYLAND_DISPLAY by a wayland / weston session.

Out of that, this is also what QT is doing. Therefore what we do is simple.

And the good, each and ever user that wants a different behaviour can export whatever he wants :-).

**Comment by bacon-cheeseburger** [comment]
@fritsch You've already back-tracked on irc and now admit that simply setting DISPLAY **does not** automatically mean an x session is running. We don't need to continue here. Kodi isn't going to use a more reliable method to determine what's running and your pride needs time to heal so we've reach the end.

**Comment by fritsch** [comment]
For the record: kodi-gbm does these checks to find out if it is running on a text-only terminal. Text-only terminals neither expose DISPLAY nor WAYLAND_DISPLAY. This is used to not trying to take over DRM-MASTER role when already a graphical user session is running. The check for the availability of DISPLAY and / or WAYLAND_DISPLAY does in no way assume, that the existence of such a variable proves a running and working Xserver / Wayland somewhere in the system.

If you find a text terminal anywhere in the wild where DISPLAY / WAYLAND_DISPLAY is hardcoded to a value without actually a graphical session is tied to this login, please file a bug with your distribution.

And as always: If someone knows a more sane way to do more "deep testing" without involving adding X11 / wayland dependencies to e.g. GBM binary, ideas are welcome.

The use-cause of those checks are as follows:
Providing a single version via our ppa for Ubuntu. In the past many new Ubuntu users installed a wayland desktop and kodi automatically used the X-Wayland X11 emulation, producing huge issues with hw decoding and OpenGL rendering.

Therefore this feature is designed, that:
if WAYLAND_DISPLAY is exposed (20.04 LTS with wayland) -> use wayland.
if DISPLAY is set but not WAYLAND_DISPLAY -> X11
If neither is set, as we run from a standard compliant text only terminal -> GBM.

Thanks for your contribution.

**Comment by bacon-cheeseburger** [comment]
@fritsch Why openly admit here that the mere existence of DISPLAY/WAYLAND_DISPLAY doesn't prove a graphical session is running, but spend the last few hours on irc ducking & dodging conceding that simple fact?

**Comment by fritsch** [comment]
There is nothing to admit. This is trivially clear for everyone, you can export whatever you want in your shell, if it's sane or not, everyone is free to do so. Though when it comes to standard shell environment variables, this has side effects.

If you just for fun export DISPLAY=whatever and then complain that X11 applications would all try to connect to "whatever", you won't really have a reason to complain. If an application honors user's environment variables (e.g. DISPLAY) by being careful and not trying to start a GBM session when found, this is another side-effect.

In short: Don't export any standard variables to improper values, if you want to have proper system behaviour.

**Comment by bacon-cheeseburger** [comment]
@fritsch I simply said "the mere existence of DISPLAY/WAYLAND_DISPLAY does not automatically mean a x/wayland session is running". For whatever reason you've been trying to argue against that fact, though other Kodi devs had no problem acknowledging its' truth. I suspect pride is the problem - not wanting to admit my statement is correct. But then you came here and said "The check for the availability of DISPLAY and / or WAYLAND_DISPLAY does in no way assume, that the existence of such a variable proves a running and working Xserver / Wayland somewhere in the system", which contradicts all your efforts on irc. Perhaps this discussion shouldn't continue until you decide which answer you want to stick with - either you think the existence of those variables does or does not mean a graphical session is running. Or maybe stop beating this dead horse.

**Comment by fritsch** [comment]
I have no idea why you try to get personal. During the conversation on IRC, no one said - that having a DISPLAY exposed means, that there is an X11 server up and running and server connections. We tried to explain you the difference between adequate and sufficient. We tried to explain you that export certain values is a problem with standard applications, but it did not help.

There is no problem with pride or something, therefore I don't care too much to be honest. Though, I won't admit to something that is not true. The IRC history is public and everyone can read, it's logged on freenode - therefore I invite everyone to make his own judgements, from my side enough is being said.
